model/hasfulltablecache.py

```diff
@@ -1,140 +1,91 @@
 # encoding: utf-8
 # HasFullTableCache
 
-import logging
+from collections import namedtuple
+from types import SimpleNamespace
+from typing import Callable, Dict, Hashable, Optional, Tuple
 
-from . import get_one
+from sqlalchemy.orm import Session
 
+from . import Base, get_one
 
-class HasFullTableCache(object):
-    """A mixin class for ORM classes that maintain an in-memory cache of
-    (hopefully) every item in the database table for performance reasons.
-    """
 
-    RESET = object()
+class HasFullTableCache(object):
+    CacheTuple = namedtuple("CacheTuple", ["id", "key", "stats"])
 
-    # You MUST define your own class-specific '_cache' and '_id_cache'
-    # variables, like so:
-    #
-    # _cache = HasFullTableCache.RESET
-    # _id_cache = HasFullTableCache.RESET
+    """
+    A mixin class for ORM classes that maintain an in-memory cache of
+    items previously requested from the database table for performance reasons.
 
-    @classmethod
-    def reset_cache(cls):
-        cls._cache = cls.RESET
-        cls._id_cache = cls.RESET
+    Items in this cache are always maintained in the same database session.
+    """
 
     def cache_key(self):
         raise NotImplementedError()
 
     @classmethod
-    def _cache_insert(cls, obj, cache, id_cache):
-        """Cache an object for later retrieval, possibly by a different
-        database session.
-        """
+    def _cache_insert(cls, obj: Base, cache: CacheTuple):
```

Comment:
I think it should be `obj: HasFullCachTable` since we're using `cache_key` below in the code.

Reply:
Updated to use a protocol, I think this addresses that concern.